### PR TITLE
Make hash collisions in the registry much less likely

### PR DESCRIPTION
**Pull request by beorn7**
Sadly, that's not yet a complete solution (like the one planned for v2, see last item in #222). However, using xxhash for the registry improves things a lot. It also moves from summing to XOR'ing for the collectorID. (Hashing the hashes would be an option, too, but it would require sorting of the hashes. I didn't want to go down that rabbit hole and rather stick with a commutative and associative operation.)

(Interestingly, just moving from sum to XOR solved #633 already, but of course, other hash collisions might happen in practice despite the XOR. In fact, hash collisions might still occur. This PR is mostly there to make this sufficiently unlikely so that the world will survive until v2 is released.)

### CODE REVIEW

**Comment by beorn7** [comment]
@awilliams this will be of interest for you.

**Comment by beorn7** [comment]
Id' consider this a fix for #633. The "proper" solution is contained in #222.

**Comment by awilliams** [comment]
Thanks @beorn7, this looks like a great bridge between now and v2. Nice to see the collision test case as well!

**Comment by beorn7** [comment]
Thanks for having a look.

File: prometheus/metric.go [hunk 1]
```diff
@@ -24,6 +24,8 @@ import (
 
 const separatorByte byte = 255
 
+var separatorByteSlice = []byte{255} // For convenient use with xxhash.
```
Comment (cstyan) [inline prometheus/metric.go]: not that it's likely to change, but should this not be `[]byte{separatorByte}`

Comment (beorn7) [reply prometheus/metric.go]: Yeah, that's better. (Originally, I wanted to remove separatorByte altogether, but then realized I want to keep the old hashing for vectors.) Will sent you a PR.

**Pull request merged**